### PR TITLE
Improve diagnostics for JarCache write errors (related to JENKINS-36947)

### PR DESCRIPTION
**Pull request by olivergondza**
https://issues.jenkins-ci.org/browse/JENKINS-36947

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
👍

File: src/main/java/hudson/remoting/Util.java [hunk 1]
```diff
@@ -247,4 +251,26 @@ static InetSocketAddress getResolvedHttpProxyAddress(String host, int port) thro
...
+    @IgnoreJRERequirement @SuppressWarnings("Since15")
+    static void mkdirs(@Nonnull File file) throws IOException {
+        if (file.isDirectory()) return;
+
+        try {
+            Class.forName("java.nio.file.Files");
+            Files.createDirectories(file.toPath());
+            return;
+        } catch (ClassNotFoundException e) {
+            // JDK6
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Util.java]: I think we can finally drop support of Java 6 in remoting.
@stephenc is going to introduce some incompatible code, and maybe we will need remoting 3 in any case

File: src/main/java/hudson/remoting/Util.java [hunk 2]
```diff
@@ -247,4 +251,26 @@ static InetSocketAddress getResolvedHttpProxyAddress(String host, int port) thro
...
+        } catch (ExceptionInInitializerError e) {
+            // JDK7 on multibyte encoding (http://bugs.java.com/bugdatabase/view_bug.do?bug_id=7050570)
+        }
+
+        // Fallback
+        if (!file.mkdirs()) {
+            if (!file.isDirectory()) {
+                throw new IOException("Directory not created");
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Util.java]: Maybe not ideal. E.g. we do not check this directory is writable, etc.
But it's definitely better than the original code

**Comment by oleg-nenashev** [comment]
re-:+1:

File: src/main/java/hudson/remoting/Util.java (hunk 1 above)
Comment (olivergondza) [inline src/main/java/hudson/remoting/Util.java]: Is there any chance to cut a JDK6 release with this fix, first?

File: src/main/java/hudson/remoting/Util.java (hunk 2 above)
Comment (olivergondza) [inline src/main/java/hudson/remoting/Util.java]: Which directory, the target one? I think that is out of the scope of the method.

File: src/main/java/hudson/remoting/Util.java (hunk 1 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Util.java]: Sure. I've forgotten you're on 1.580.
Right now there is no strong need to migrate, so I would not like to hurry too much

Comment (olivergondza) [inline src/main/java/hudson/remoting/Util.java]: Not any more, it is 1.609 but the limitation remains.

File: src/main/java/hudson/remoting/Util.java (hunk 2 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Util.java]: Maybe

**Pull request merged**